### PR TITLE
Bremsstrahlung radiation source

### PR DESCRIPTION
**Pull request by allen-adastra**
Introduce a Bremmstrahlung radiation source.

- `BremsstrahlungHeatSink`
- User can specify whether to use a relativistic correction due to Stott 2005 PPCF
- @araju fixed a bug in `torax/sources/source.py`

### CODE REVIEW

File: torax/sources/radiation_source.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        dynamic_runtime_params_slice.numerics.nref,
+    )
+    # As a sink, the power is negative.
+    return -1.0 * P_brem_profile
+
+
+@dataclasses.dataclass(kw_only=True)
+class BremsstrahlungHeatSink(source.SingleProfileTempElSource):
+  """Fusion heat source for both ion and electron heat."""
+
```
[the remark this answers is not in the document]
Comment (jcitrin) [reply torax/sources/radiation_source.py:75]: This raises a typeerror bug but with the help of @araju we can fix it

**Review by jcitrin — COMMENTED** [review]
In addition to this, would be good to add Qbrem output (the radial profile) to the core_sources output object

File: torax/sources/radiation_source.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  # model_func: source.SourceProfileFunction = bremsstrahlung_model_func
```
Comment (allen-adastra) [inline torax/sources/radiation_source.py:76]: When I use the type annotation, I get this error @araju :
```
(toraxenv) awang@mfews-awang:~/repos/torax (allenw/bremss_radiation)$ python3 -m pytest torax/sources/tests/radiation_source.py --pdb
========================================================================== test session starts ===========================================================================
platform linux -- Python 3.10.14, pytest-8.2.2, pluggy-1.5.0
rootdir: /home/awang/repos/torax
configfile: pyproject.toml
plugins: typeguard-2.13.3, jaxtyping-0.2.30
collected 6 items                                                                                                                                                        

torax/sources/tests/radiation_source.py ....F
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

source_init_kwargs = {'affected_core_profiles': (<AffectedCoreProfile.TEMP_EL: 4>,), 'formula': None, 'model_func': <function bremsstrahlung_model_func at 0x75161a5ba200>, 'supported_modes': (<Mode.ZERO: 0>, <Mode.FORMULA_BASED: 2>)}
context_msg = 'making builder'

    def check_kwargs(source_init_kwargs, context_msg):
      for f in source_fields:
        v = source_init_kwargs[f.name]
        if isinstance(f.type, str):
          if f.type == 'tuple[AffectedCoreProfile, ...]':
            assert isinstance(v, tuple)
            assert all([isinstance(var, AffectedCoreProfile) for var in v])
          elif f.type == 'tuple[runtime_params_lib.Mode, ...]':
            assert isinstance(v, tuple)
            assert all([isinstance(var, runtime_params_lib.Mode) for var in v])
          elif f.type == 'SourceProfileFunction | None':
            assert v is None or callable(v)
          elif f.type == 'source.SourceProfileFunction':
            if not callable(v):
              raise TypeError(
                  f'While {context_msg} {source_type} got field '
                  f'{f.name} of type source.SoureProfileFunction '
                  ' but was passed constructor argument with value '
                  f'{v} of type {type(v)}. It is not callable, so '
                  'it cannot be a SourceProfileFunction.'
              )
          elif f.type in [
              'source.SourceOutputShapeFunction',
              'SourceOutputShapeFunction',
          ]:
            if not callable(v):
              raise TypeError(
                  f'While {context_msg} {source_type} got field '
                  f'{f.name} of type source.SoureProfileFunction '
                  ' but was passed constructor argument with value '
                  f'{v} of type {type(v)}. It is not callable, so '
                  'it cannot be a SourceProfileFunction.'
              )
          else:
            raise TypeError(f'Unrecognized type string: {f.type}')
        else:
          try:
>           type_works = isinstance(v, f.type)

torax/sources/source.py:758: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.pyenv/versions/3.10.14/lib/python3.10/typing.py:994: in __instancecheck__
    return self.__subclasscheck__(type(obj))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = typing.Callable[[torax.config.runtime_params_slice.DynamicRuntimeParamsSlice, torax.sources.runtime_params.DynamicRunt...mpy.bool, numpy.number, typing.Iterable[ForwardRef('ArrayTree')], typing.Mapping[typing.Any, ForwardRef('ArrayTree')]]]
cls = <class 'function'>

    def __subclasscheck__(self, cls):
>       raise TypeError("Subscripted generics cannot be used with"
                        " class and instance checks")
E       TypeError: Subscripted generics cannot be used with class and instance checks

../../.pyenv/versions/3.10.14/lib/python3.10/typing.py:997: TypeError

The above exception was the direct cause of the following exception:

self = <radiation_source.BremsstrahlungHeatSinkTest testMethod=test_invalid_source_types_raise_errors>

    def test_invalid_source_types_raise_errors(self):
      """Tests that using unsupported types raises an error."""
      runtime_params = general_runtime_params.GeneralRuntimeParams()
      geo = geometry.build_circular_geometry()
      # pylint: disable=missing-kwoa
>     source_builder = self._source_class_builder()  # pytype: disable=missing-parameter

torax/sources/tests/test_lib.py:124: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
<string>:8: in __init__
    ???
torax/sources/source.py:776: in __post_init__
    check_kwargs(source_init_kwargs, 'making builder')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

source_init_kwargs = {'affected_core_profiles': (<AffectedCoreProfile.TEMP_EL: 4>,), 'formula': None, 'model_func': <function bremsstrahlung_model_func at 0x75161a5ba200>, 'supported_modes': (<Mode.ZERO: 0>, <Mode.FORMULA_BASED: 2>)}
context_msg = 'making builder'

    def check_kwargs(source_init_kwargs, context_msg):
      for f in source_fields:
        v = source_init_kwargs[f.name]
        if isinstance(f.type, str):
          if f.type == 'tuple[AffectedCoreProfile, ...]':
            assert isinstance(v, tuple)
            assert all([isinstance(var, AffectedCoreProfile) for var in v])
          elif f.type == 'tuple[runtime_params_lib.Mode, ...]':
            assert isinstance(v, tuple)
            assert all([isinstance(var, runtime_params_lib.Mode) for var in v])
          elif f.type == 'SourceProfileFunction | None':
            assert v is None or callable(v)
          elif f.type == 'source.SourceProfileFunction':
            if not callable(v):
              raise TypeError(
                  f'While {context_msg} {source_type} got field '
                  f'{f.name} of type source.SoureProfileFunction '
                  ' but was passed constructor argument with value '
                  f'{v} of type {type(v)}. It is not callable, so '
                  'it cannot be a SourceProfileFunction.'
              )
          elif f.type in [
              'source.SourceOutputShapeFunction',
              'SourceOutputShapeFunction',
          ]:
            if not callable(v):
              raise TypeError(
                  f'While {context_msg} {source_type} got field '
                  f'{f.name} of type source.SoureProfileFunction '
                  ' but was passed constructor argument with value '
                  f'{v} of type {type(v)}. It is not callable, so '
                  'it cannot be a SourceProfileFunction.'
              )
          else:
            raise TypeError(f'Unrecognized type string: {f.type}')
        else:
          try:
            type_works = isinstance(v, f.type)
          except TypeError as exc:
>           raise TypeError(
                f'While {context_msg} {source_type} got field '
                f'{f.name} whose type is {f.type} of type'
                f'{type(f.type)}. This is not a valid type.'
            ) from exc
E           TypeError: While making builder <class 'torax.sources.radiation_source.BremsstrahlungHeatSink'> got field model_func whose type is typing.Callable[[torax.config.runtime_params_slice.DynamicRuntimeParamsSlice, torax.sources.runtime_params.DynamicRuntimeParams, torax.geometry.Geometry, torax.state.CoreProfiles | None], typing.Union[jax.Array, numpy.ndarray, numpy.bool, numpy.number, typing.Iterable[ForwardRef('ArrayTree')], typing.Mapping[typing.Any, ForwardRef('ArrayTree')]]] of type<class 'typing._CallableGenericAlias'>. This is not a valid type.

torax/sources/source.py:760: TypeError
```

**Comment by allen-adastra** [comment]
@araju  also getting this error.

```

torax/sources/tests/bremsstrahlung_heat_sink.py ....F
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

self = <bremsstrahlung_heat_sink.BremsstrahlungHeatSinkTest testMethod=test_invalid_source_types_raise_errors>

    def test_invalid_source_types_raise_errors(self):
      """Tests that using unsupported types raises an error."""
      runtime_params = general_runtime_params.GeneralRuntimeParams()
      geo = geometry.build_circular_geometry()
      # pylint: disable=missing-kwoa
      source_builder = self._source_class_builder()  # pytype: disable=missing-parameter
      # pylint: enable=missing-kwoa
      source_models_builder = source_models_lib.SourceModelsBuilder(
          {'foo': source_builder},
      )
      source_models = source_models_builder()
      source = source_models.sources['foo']
      self.assertIsInstance(source, source_lib.SingleProfileSource)
      dynamic_runtime_params_slice = (
          runtime_params_slice.build_dynamic_runtime_params_slice(
              runtime_params=runtime_params,
              sources=source_models_builder.runtime_params,
          )
      )
      core_profiles = core_profile_setters.initial_core_profiles(
          dynamic_runtime_params_slice=dynamic_runtime_params_slice,
          geo=geo,
          source_models=source_models,
      )
      for unsupported_mode in self._unsupported_modes:
        source_builder.runtime_params.mode = unsupported_mode
        dynamic_runtime_params_slice = (
            runtime_params_slice.build_dynamic_runtime_params_slice(
                runtime_params=runtime_params,
                sources=source_models_builder.runtime_params,
            )
        )
        with self.subTest(unsupported_mode.name):
>         with self.assertRaises(jax.interpreters.xla.xe.XlaRuntimeError):
E         AssertionError: XlaRuntimeError not raised

torax/sources/tests/test_lib.py:152: AssertionError
```

File: torax/sources/radiation_source.py
[the remark this answers is not in the document]
Comment (allen-adastra) [reply torax/sources/radiation_source.py:1]: Done

File: torax/sources/radiation_source.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        core_profiles,
+        dynamic_runtime_params_slice.runtime_params.plasma_composition.Zeff,
+        dynamic_runtime_params_slice.numerics.nref,
+    )
+    # As a sink, the power is negative.
+    return -1.0 * P_brem_profile
+
+
+@dataclasses.dataclass(kw_only=True)
+class BremsstrahlungHeatSink(source.SingleProfileTempElSource):
```
[the remark this answers is not in the document]
Comment (allen-adastra) [reply torax/sources/radiation_source.py]: Done

**Comment by allen-adastra** [comment]
> core_sources

Seems its already automatically added to `SourceProfiles.profiles` or did you have something else in mind?

**Pull request merged**